### PR TITLE
[Feature 8] Enemy System

### PR DESCRIPTION
**Pull request by Yasutakahi**
**implemented enemy system + doc string + tests**

**1. Enemy entities**

- EnemyInfo: EnemyInfo class has all the information about the enemy including health, reputation, speed, skills, and type. The class has getter and setter methods. Since there was a bidirectional dependency between EnemyAI and Enemy, this class was created to avoid that dependency.

- Enemy: Enemy is a subclass of character. This class has name, EnemyInfo, and EnemyAI. This class also has setter and getter functions.

- Boss: Boss extends enemy class. This class basically has same instance attributes as enemy and another attributes called gimmick. Boss has a method called useGimmick which calls a a method in Gimmick class to do gimmick. Also, it has another instance attributes called 

**2. Enemy AI**

- EnemyAI: EnemyAI is an interface for enemy AI. The enemy class depends on this EnemyAI interface for dependency inversion.

- DefaultAI: DefaultAI implements the EnemyAI class. The constructor of a Default AI class has input and chance of attack as parameters. Also it has method called respond which returns enemy’s action depending on the chance of attack and user’s action.

- SmartAI: Smart AI has same method but the implementation is different. It gives more reasonable action. For example, when the enemy’s health is low, it will heal more likely and in the same way, when the health is high, it will more likely attack the user.

**3. Gimmick**

- Gimmick: Like EnemyAI, Enemy class depends on this interface for the dependency inversion. This class takes EnemyInfo as a parameter (constructor). It has gimmickCheck which checks if the gimmick is triggered or not and useGimmick which uses gimmick (updating stats). It updates the EnemyInfo so it does not directly access to the Enemy object. Some gimmicks are triggered more easily than other gimmicks

- AttackGimmick: This gimmick increases the damage of the skills that the enemy has by 1.2. Since this gimmick is more powerful than other gimmicks, it is slightly hard to trigger.

- HealthGimmick: This gimmick fully heals enemy’s health when it is triggered. It is slightly easy to trigger

- SpeedGimmick: SpeedGimmick increases the enemy’s speed so that the enemy can takes the first turn more likely.

- TypeGimmick: TypeGimmick changes the type of the enemy. It can be beneficial but sometimes not, so it is relatively easier to trigger.

**4. Enemy Factory**

- EnemyFactory: It takes an information stored in a database and create enemy and boss. There is methods called getEnemy and getBoss which returns the enemy and boss objects respectively.

- EnemyAIFactory: It creates an enemyAI depending on the enemy with the specific name has. For example, if the name of the enemy is goblin, it goes to the database to find the enemy called goblin and return the enemyAI that goblin has. The default AI is set to DefaultAI.

- EnemyInfoFactory: It basically creates EnemyInfo instance from the database with the given name.

- GimmickNotFoundException and TypeNotFoundException: those are the exceptions throw when the gimmick and type in the database are not appropriate gimmick and type respectively.

### CODE REVIEW

File: src/main/java/objects/battle/enemy/EnemyFactory.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        EnemyAI AItype;
+        String ai = enemyData.ai;
+        AIDataManager dataAI = new AIDataManager();
+        AIData aiData = dataAI.fetchAIData(ai);
+        if (Objects.equals(ai, "smart")) {
+            AItype = new SmartAI(enemyInfo, Integer.parseInt(aiData.chance));
+        } else{
+            AItype = new DefaultAI(enemyInfo, Integer.parseInt(aiData.chance));
+        }
+        return new Enemy(name, enemyInfo, AItype);
```
Comment (bw55555) [inline src/main/java/objects/battle/enemy/EnemyFactory.java:61]: why use name here? should be based on the display name of the enemy from the database
likely you will be passed the id of the enemy into this method which could be different from the name to be displayed to the user

**Review by bw55555 — CHANGES_REQUESTED** [review]
Looks like good test coverage and understandable docstring

Some problems with clean architecture that might need to be addressed

**Review by charlietaoxl — DISMISSED** [review]
So far the docstrings and unit tests all look fleshed out. 

A couple of things that can be added in the future could be:
baseExp attribute
getMaxHealth() method
baseSpeed attribute (since speed changes during a battle, this is useful when resetting)
resetSpeed() method

In the future, we might need a BattleCharacter class that can include Enemy and Player battle information to standardize the code, but that can be discussed later. For now this is fine!

File: src/main/java/objects/battle/enemy/EnemyInfo.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    /**
+     * This function sets the enemy's health
+     *
+     * @param n: new health in int
+     */
+    public void setHealth(int n)
+    {
+        this.health = n;
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/battle/enemy/EnemyInfo.java:93]: Consider renaming the parameter variables for clarity.

File: src/main/java/objects/battle/enemy/EnemyInfo.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    /**
+     * This method returns the skill at index n
+     *
+     * @param n: the index of the skill
+     * @return the skill at index n
+     */
+    public Skill getSkill(int n){
+        return this.skills.get(n);
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/battle/enemy/EnemyInfo.java:123]: Rename parameter names to something like index.

File: src/main/java/objects/battle/enemy/factory/EnemyFactory.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+     * This method returns the enemy or boss instance using the information given by database
+     * This method returns Enemy so if the boss is returned, make sure you cast, so that the object
+     * is Boss.
+     * @param name of the enemy to create
+     * @return instance of enemy
+     */
+    public static Enemy createEnemy(String name) throws Exception {
+        EnemyInfo enemyInfo = EnemyInfoFactory.createEnemyInfo(name);
+        EnemyAI enemyAI = EnemyAIFactory.createEnemyAI(name);
+        EnemyDataManager data = new EnemyDataManager();
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/battle/enemy/factory/EnemyFactory.java:26]: You should not be creating database managers everytime you call a method. Instead, create a private instance of a database at the highest level of the class hierarchy and pass the database through as a parameter when calling a method.

File: src/main/java/objects/battle/enemy/factory/EnemyInfoFactory.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+     * This method returns the EnemyInfo that the enemy with the name in the database has.
+     * @param name of the enemy
+     * @return EnemyInfo of the enemy with the name given
+     */
+    public static EnemyInfo createEnemyInfo(String name) throws TypeNotFoundException {
+        EnemyDataManager data = new EnemyDataManager();
+        EnemyData enemyData = data.fetchEnemyData(name);
+        ArrayList<Skill> skills = new ArrayList<>();
+        SkillDataManager dataSkill = new SkillDataManager();
+        for (String i:enemyData.skills) {
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/battle/enemy/factory/EnemyInfoFactory.java:25]: Consider spacing code out more for clarity. Comments would be appreciated.

File: src/main/java/objects/battle/enemy/factory/EnemyFactory.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        EnemyData enemyData = data.fetchEnemyData(name);
+        if(!(enemyData.gimmick ==null)){
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/battle/enemy/factory/EnemyFactory.java:28]: Try enemyData.gimmick != null

File: src/main/java/objects/battle/enemy/factory/EnemyAIFactory.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * This method returns the Enemy AI that the enemy with the name in the database has.
+     * @param name of the enemy to create
+     * @return instance enemy AI tha the enemy with the name has
+     */
+    public static EnemyAI createEnemyAI(String name) throws TypeNotFoundException {
+        EnemyDataManager data = new EnemyDataManager();
+        EnemyData enemyData = data.fetchEnemyData(name);
+        EnemyInfo enemyInfo = EnemyInfoFactory.createEnemyInfo(name);
+        EnemyAI AItype;
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/battle/enemy/factory/EnemyAIFactory.java:25]: Consider spacing out code for neatness.

File: src/main/java/objects/character/Boss.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        this.gimmick = gimmick;
+    }
+
+    /**
+     * This method returns the enemy's gimmick
+     *
+     */
+    public Gimmick getGimmick() {
+        return gimmick;
+    }
```
Comment (mahirhassan-uoft) [inline src/main/java/objects/character/Boss.java:85]: Gimmicks system is very interesting, good work!

File: src/test/java/battle/enemy/EnemyFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        Assertions.assertEquals(boss.getReputation(), 20);
+        Assertions.assertEquals(boss.getSpeed(), 110);
+        Assertions.assertEquals(boss.getType(), SkillType.EARTH);
+        Assertions.assertEquals(boss.getHealth(), 100);
+        Assertions.assertEquals(boss.getName(), "goblin warrior");
+        Assertions.assertTrue(boss.getGimmick() instanceof
+                HealthGimmick);
+        Assertions.assertTrue(boss.getEnemyAI() instanceof
+                SmartAI);
+    }
```
Comment (mahirhassan-uoft) [inline src/test/java/battle/enemy/EnemyFactoryTest.java:67]: Many tests are used for functional purposes, very nice.

File: src/test/java/battle/enemy/EnemyInfoTest.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        ArrayList<Skill> skills = new ArrayList<Skill>();
+        skills.add(skill);
+        EnemyInfo enemyInfo = new EnemyInfo(skills, 90, 10, SkillType.WATER);
+        Assert.assertEquals(enemyInfo.getType(), SkillType.WATER);
+    }
+
+    @DisplayName("Test for setType method")
+    @Test
+    public void setTypeTest(){
+        Skill skill = new Skill("fire ball", 20, 5, SkillType.WATER);
```
Comment (mahirhassan-uoft) [inline src/test/java/battle/enemy/EnemyInfoTest.java:132]: Small nitpick, but maybe rename "fire ball" to "fireball"

**Review by mahirhassan-uoft — CHANGES_REQUESTED** [review]
Nice work, however you need to move databases to the top of the class hierarchy instead of creating new instances of it everytime you call a method. After these changes, your code will be both functional and effective. Other than this, nice work :)

**Review by mahirhassan-uoft — APPROVED** [review]
nice changes

**Review by charlietaoxl — APPROVED** [review]
Fixes were made where needed, looking up to date and good for dependencies!

**Review by bw55555 — APPROVED** [review]
Looks good now!

**Pull request merged**